### PR TITLE
[StructuralMechanicsApplication] Hotfix MPC builder and solver

### PR DESCRIPTION
**Pull request by loumalouomega**
Partially solves #1812. It solves the problem only for the MPC bullder and solver

Apparently after #1468 we must be very careful with the keys. So must use std::size_t or std::uint64_t

Changes:
- I re-factorized the MPC builder and solver @adityaghantasala checks eveything is right, test passes. I reduce code duplication creating templated methods for conditions and elements. I replaced all int and unsigned int for std::size_t. I added the logger and I apply the format (not 100%, but in all the code i modified)
- Idem for the MPC proces. I replaced the new for make_shared and I both check Has and if the value is nullptr
- Idem for the MPC data. I replaced the unsigned int for std::size_t. Besides I replaces boost hash for the Kratos HashCombine that was created for the mmg_process

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h [hunk 1]
```diff
@@ -32,134 +29,206 @@
...
-                        if (KratosComponents<Variable<double>>::Has(varName))
-                        { //case of double variable
-                            VariableType rVar = KratosComponents<Variable<double>>::Get(m_parameters["variable_names"][i].GetString());
+            Element::Pointer p_master_element;
+            const bool is_found = p_point_locator->FindPointOnMesh(p_slave_node->Coordinates(), N, p_master_element, result_begin, max_results);
+            if (is_found == true) {
+                for (IndexType i = 0; i < num_vars; i++) {
+                    std::string varName = mParameters["variable_names"][i].GetString();
+                    Geometry<NodeType> &geom = p_master_element->GetGeometry();
+                    for (IndexType i = 0; i < geom.size(); i++) {
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:226]: again i??

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h [hunk 2]
```diff
@@ -32,134 +29,206 @@
...
-                    for (unsigned int i = 0; i < geom.size(); i++)
-                    {
-                        if (KratosComponents<Variable<double>>::Has(varName))
-                        { //case of double variable
-                            VariableType rVar = KratosComponents<Variable<double>>::Get(m_parameters["variable_names"][i].GetString());
+            Element::Pointer p_master_element;
+            const bool is_found = p_point_locator->FindPointOnMesh(p_slave_node->Coordinates(), N, p_master_element, result_begin, max_results);
+            if (is_found == true) {
+                for (IndexType i = 0; i < num_vars; i++) {
+                    std::string varName = mParameters["variable_names"][i].GetString();
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:224]: this is a string access within a node loop?? to parameters??

absolutely to be avoided...

**Review by RiccardoRossi — COMMENTED** [review]
I have the impression this is a very old version.

@adityaghantasala can u confirm?

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h (hunk 1 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:226]: Not my original code...

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:224]: Not my original code...

**Comment by loumalouomega** [comment]
This is the version in the master in the StructuralMechanicsApplication

**Comment by adityaghantasala** [comment]
@loumalouomega thanks for changing this. I will go through the changes in the next hours. 
@RiccardoRossi this is the version on the head. The other implementation where I made these formatting changes is on the MPC to core branch which I hoped to merge. So I did not make them again as this will be double work.

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h (hunk 2 above)
Comment (adityaghantasala) [reply applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:224]: @RiccardoRossi this means one has to loop over all the nodes on slave modelpart for each variable. instead I choose to loop over the nodes once and check using this string access. 
Do you suggest the other way is efficient ?

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h (hunk 1 above)
Comment (adityaghantasala) [reply applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:226]: I will change this in a commit to this branch, and we can merge this together with this change. But I am surprised that it did not have any problem in the cases I used.

File: applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h (hunk 2 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_processes/apply_multi_point_constraints_process.h:224]: Hi @adityaghantasala string manipulation/comparison is extremely slow and should be avoided as much as possible.

also the "Parameters" objects and the "KratosComponents" are designed more to catch errors than for fast performance. 

Long story short, i would definitely invert the loops (that is, do the variable loop as outer).

Just for curiosity time it. I am very convinced you'll see a 1-2 order of magnitude speedup.

**Comment by RiccardoRossi** [comment]
@adityaghantasala unfortunately this code will be very affected by the change of adding the MPCs to the modelpart. 

This also means that it doesn't make sense to spend a lot of time on this before actually making such changes. Long story short, most probably it simply makes sense to have it to compile and wait for updating it at a later stage

**Comment by pooyan-dadvand** [comment]
Just to add that the key type should only be `std::uint64_t` and **NOT**` std::size_t`

**Comment by loumalouomega** [comment]
What do we do with this?

**Review by adityaghantasala — APPROVED** [review]
As @RiccardoRossi  suggested, since this soon will change, I am not changing some of the changes suggested.

**Comment by adityaghantasala** [comment]
We should merge this as some people are already using the constraints and having problems (seg faults) because size_t and unsigned int.

**Pull request merged**